File: C++/findContinuousSequence.cpp
/**
 * @File    : findContinuousSequence.cpp
 * @Brief   : 和为s的连续正数序列   
 * @Link    : https://leetcode-cn.com/problems/he-wei-sde-lian-xu-zheng-shu-xu-lie-lcof/
 * @Author  : Wei Li
 * @Date    : 2021-09-25
*/

#include <iostream>
#include <vector>

class Solution
{
public:
    std::vector<std::vector<int>> findContinuousSequence(int target)
    {
        std::vector<std::vector<int>> vec;
        std::vector<int> res;
        for (int l = 1, r = 2; l < r;)
        {
            int sum = (l + r) * (r - l + 1) / 2;
            if (sum == target)
            {
                res.clear();
                for (int i = l; i <= r; ++i)
                {
                    res.emplace_back(i);
                }
                vec.emplace_back(res);
                l++;
            }
            else if (sum < target)
            {
                r++;
            }
            else
            {
                l++;
            }
        }
        return vec;
    }
};

template <typename T>
void DisplayElement(const T &container)
{
    for (auto element : container)
    {
        for (auto value : element)
        {
            std::cout << value << " ";
        }
        std::cout << std::endl;
    }
    std::cout << std::endl;
}

// -----------------------------
int main(int argc, char **argv)
{
    int target = 9;

    Solution solution;
    auto find_continuous_sequence = solution.findContinuousSequence(target);
    std::cout << "The solution of this problem is " << std::endl;
    DisplayElement(find_continuous_sequence);

    return 0;
}

File: Python/findContinuousSequence.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: findContinuousSequence.py
@Function: 和为s的连续正数序列
@Link: https://leetcode-cn.com/problems/he-wei-sde-lian-xu-zheng-shu-xu-lie-lcof/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-09-25
"""


from typing import List


class Solution:
    def findContinuousSequence(self, target: int) -> List[List[int]]:
        i, j, s, res = 1, 2, 3, []
        while i < j:
            if s == target:
                res.append(list(range(i, j + 1)))
            if s >= target:
                s -= i
                i += 1
            else:
                j += 1
                s += j
        return res


# --------------------------------------
if __name__ == "__main__":
    number = 9

    solution = Solution()
    find_continuous_sequence = solution.findContinuousSequence(number)
    print(f"The solution of this problem is \n{find_continuous_sequence}")
